Speed up import ~9x by using one shared Lark parser

agentydragon opened the pull request:
Previously, make_syntax_validator() created a new Lark parser for each rule, which was expensive (~660ms total import time). This change:

- Defines ALL_START_RULES containing all grammar rules needed by validators
- Creates a single shared syntax_parser with all start rules upfront
- Refactors make_syntax_validator() to use functools.partial with is_valid_syntax() instead of creating new parsers

Import time reduced from ~660ms to ~72ms (9x speedup).

Also fixed a forgotten comma that previously caused implicit string literal concatenation:

```python
    "irelative_part"
    "ihier_part",
```

agentydragon committed:
Optimize import time by using a single shared Lark parser

Previously, make_syntax_validator() created a new Lark parser for each
rule, which was expensive (~660ms total import time). This change:

- Defines ALL_START_RULES containing all grammar rules needed by validators
- Creates a single shared syntax_parser with all start rules upfront
- Refactors make_syntax_validator() to use functools.partial with
  is_valid_syntax() instead of creating new parsers

Import time reduced from ~660ms to ~72ms (9x speedup).

Testing protocol:
- Created timing script that clears module cache and measures import time
- Ran 5 iterations for each version to account for variance
- Used git stash to preserve dirty state while testing HEAD
- Results: HEAD averaged ~662ms, this change averages ~72ms
- All existing tests pass (2/2)

agentydragon commented:
@willynilly hi! do you think we could merge this? or would you like me to do any changes / testing?

paugier commented:
Shouldn't these changes be merged? Are there any maintainers active here?

jkowalleck commented:
> Shouldn't these changes be merged? Are there any maintainers active here?

see https://github.com/willynilly/rfc3987-syntax/issues/15

agentydragon committed:
Give the fork a local version so lockfiles can tell it apart

The fork declared version 1.1.0, exactly the upstream release it patches. That is
indistinguishable from PyPI's 1.1.0 to a resolver, and it silently defeats the fork
downstream: uv resolves a URL/git requirement for this package, matches name+version
against the registry, and emits the *registry's* hashes for it. A consumer generating a
hashed lockfile then either installs upstream anyway or fails the checksum.

That is what happened in ducktape, which has carried a dead override for this fork since
2025-12-30 without the optimisation ever being in effect.

1.1.0+adgn.1 is a PEP 440 local version: it compares equal to 1.1.0 for `==1.1.0` and
satisfies `>=1.1.0`, so nothing that depends on the upstream release stops resolving —
but the version string is no longer one PyPI serves, so hashes cannot be borrowed.
